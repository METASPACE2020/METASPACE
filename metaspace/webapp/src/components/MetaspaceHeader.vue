<template>
  <div class="b-header">
    <div class="header-items">
      <router-link to="/" class="header-item logo">
        <img src="../assets/logo.png" alt="Metaspace" title="Metaspace"/>
      </router-link>

      <router-link :to="uploadHref" class="header-item page-link" id='upload-link'>
        Upload
      </router-link>

      <router-link :to="datasetsHref" class="header-item page-link" id='datasets-link'>
        Datasets
      </router-link>

      <router-link :to="annotationsHref" class="header-item page-link" id='annotations-link'>
        Annotations
      </router-link>

      <router-link to="/about" class="header-item page-link">
        About
      </router-link>

      <router-link to="/help" class="header-item page-link">
        Help
      </router-link>
    </div>

    <div v-if="!this.$store.state.authenticated" class="header-items">
      <div v-if="features.newAuth" class="header-items">
        <div class="header-item page-link" @click="showCreateAccount">
          Create account
        </div>

        <div class="header-item page-link" @click="showSignIn">
          Sign in
        </div>
      </div>
      <div v-else class="header-items">
        <el-popover placement="bottom" trigger="click" class="header-items">
          <div slot="reference" class="header-item page-link">
            Sign in
          </div>
          <div id="email-link-container">
            <el-button type="primary" @click="sendLoginLink">Send a link to</el-button>
            <span>
              <el-input v-model="loginEmail" placeholder="e-mail address" />
            </span>
          </div>

          <div style="text-align: center;">
            <div style="margin: 10px; font-size: 18px;">or</div>
            <a href="/auth/google">
              <el-button>Sign in with Google</el-button>
            </a>
          </div>
        </el-popover>
      </div>
    </div>
    <div v-else class="header-items">
      <router-link
        v-if="currentUser && currentUser.primaryGroup"
        :to="`/group/${currentUser.primaryGroup.group.id}`"
        class="header-item page-link">
        <div class="limit-width">
          {{currentUser.primaryGroup.group.shortName}}
        </div>
      </router-link>
      <div class="submenu-container user-submenu"
           :class="{'submenu-container-open': openSubmenu === 'user'}"
           @mouseenter="handleSubmenuEnter('user')"
           @mouseleave="handleSubmenuLeave('user')">
        <div class="header-item submenu-header"
             :class="{'router-link-active': matchesRoute('/user/me')}">
          <div class="limit-width" style="color: white;">
          {{ userNameOrEmail }}
          </div>
        </div>
        <div class="submenu">
          <router-link to="/user/me" class="submenu-item page-link">
            My account
          </router-link>
          <div class="submenu-item page-link" @click="logout">
            Sign out
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
  import gql from 'graphql-tag';
 import {signOut} from '../api/auth';
 import {encodeParams} from '../url';
  import { refreshLoginStatus } from '../graphqlClient';
 import * as config from '../clientConfig.json';

 export default {
   name: 'metaspace-header',

   computed: {
     uploadHref() {
       return this.href('/upload');
     },

     datasetsHref() {
       return this.href('/datasets');
     },

     annotationsHref() {
       return this.href('/annotations');
     },

     userNameOrEmail() {
       const {user} = this.$store.state;
       if (!user)
         return '';
       return user.name || user.email;
     },

     features() {
       return config.features;
     }
   },

   data() {
     return {
       loginEmail: (this.$store.state.user ? this.$store.state.user.email : ''),
       currentUser: null,
       openSubmenu: null
     };
   },

   apollo: {
     currentUser: {
       query: gql`query {
         currentUser {
           primaryGroup {
             group {
               id
               shortName
               name
             }
           }
         }
       }`
     }
   },

   methods: {
     href(path) {
       const lastParams = this.$store.state.lastUsedFilters[path];
       let f = lastParams ? lastParams.filter : {}
       f = Object.assign({}, f, this.$store.getters.filter)
       const link = {
         path,
         query: encodeParams(f, path, this.$store.state.filterLists)
       };
       return link;
     },

     matchesRoute(path) {
       // WORKAROUND: vue-router hides its util function "isIncludedRoute", which would be perfect here
       // return isIncludedRoute(this.$route, path);
       return this.$route.path.startsWith(path);
     },

     sendLoginLink() {
       fetch('/sendToken?user=' + this.loginEmail)
         .then((res) => {
           if (res.ok)
             this.$notify({
               title: 'Check your mailbox!',
               type: 'success',
               message: "We've sent you an e-mail with the link to log in"
             });
           else
             this.$notify({
               title: 'Error ' + res.status,
               type: 'error',
               message: res.statusText
             });
         });
       this.$ga.event({
         eventCategory: 'Link sender',
         eventAction: 'sending login link'
       })
     },

     showCreateAccount() {
       this.$store.commit('account/showDialog', 'createAccount');
     },

     showSignIn() {
       this.$store.commit('account/showDialog', 'signIn');
     },

     async logout() {
       if (config.features.newAuth) {
         await signOut();
       } else {
         await fetch('/logout', {credentials: 'include'});
       }
       await refreshLoginStatus();
     },

     handleSubmenuEnter(submenu) {
       this.openSubmenu = submenu;
     },
     handleSubmenuLeave(submenu) {
       if (this.openSubmenu === submenu) {
         this.openSubmenu = null;
       }
     },
   }
 }
</script>

<style scoped>
 .b-header {
   background-color: rgba(0, 105, 224, 0.85);
   position: fixed;
   z-index: 1000;
   top: 0;
   left: 0;
   right: 0;
   height: 62px;
   display: flex;
   align-items: center;
   justify-content: space-between;
 }

 .header-items {
   display: flex;
   align-items: center;
   height: 100%
 }

 .header-item {
   display: flex;
   border: none;
   padding: 0px 20px;
   font-size: 16px;
   align-self: stretch;
   align-items: center;
   justify-content: center;
 }
 .header-item.logo {
   padding-left: 15px;
 }

 @media (max-width: 1000px) {
   .header-item {
     padding: 0px 10px;
     font-size: 14px;
   }
 }

 .page-link {
   text-align: center;
   color: #eee;
   cursor: pointer;
   text-decoration: none;
 }

 .router-link-active.page-link, .page-link:hover,
 .submenu-container:not(.submenu-container-open) > .router-link-active.submenu-header {
   background: rgba(0, 0, 0, 0.1);
   outline-offset: -1px;
   outline-color: rgba(0, 0, 0, 0.3);
   outline-style: solid;
   outline-width: 1px;
   color: white;
 }

 .router-link-active.page-link {
   font-weight: 700;
 }
 .page-link a {
   text-decoration: none;
 }

 .submenu-container {
   position: relative;
   display: flex;
   align-items: center;
   height: 100%
 }
 .submenu {
   display: none;
   position: absolute;
   flex-direction: column;
   top: 100%;
   right: 0;
   width: 100%;
   min-width: 140px;
   max-width: 200px;

   background-color: rgba(0, 105, 224, 0.85);
   /*background-color: rgb(38, 128, 229);*/
 }
 .submenu-container-open > .submenu {
   display: flex;
 }
 .submenu-item {
   padding: 20px;
   font-size: 16px;
   align-self: stretch;
   justify-content: center;
 }
 .limit-width {
   max-width: 250px;
   overflow-wrap: break-word;
   text-align: center;
   overflow: hidden;
   line-height: 1.2em;
   max-height: 2.4em;
 }
 @media (max-width: 1000px) {
   .limit-width {
     max-width: 150px;
   }
 }

 #email-link-container {
   display: inline-flex;
 }
</style>
