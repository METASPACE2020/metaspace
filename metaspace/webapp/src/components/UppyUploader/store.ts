// based on https://uppy.io/docs/stores/#Implementing-Stores

import { ref } from 'vue'
import { State } from '@uppy/core'

type Listener = (prevState: State, nextState: State, patch: object) => void

export default () => {
  const state = ref<State | any>({})
  const listeners = new Set<Listener>()

  return {
    getState: () => state.value,
    setState: (patch: object) => {
      const prevState = state.value
      const nextState = { ...prevState, ...patch }

      state.value = nextState

      listeners.forEach((listener) => {
        listener(prevState, nextState, patch)
      })
    },
    subscribe: (listener: Listener) => {
      listeners.add(listener)
      return () => listeners.delete(listener)
    },
  }
}
