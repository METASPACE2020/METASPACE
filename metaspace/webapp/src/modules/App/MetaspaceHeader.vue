<template>
  <div>
    <div :class="healthMessage ? 'spacerWithAlert' : 'spacer'" />
    <div class="b-header">
      <div class="header-items">
        <router-link to="/" class="header-item logo">
          <img src="../../assets/logo.png" alt="Metaspace" title="Metaspace"/>
        </router-link>

        <router-link :to="uploadHref" class="header-item page-link" id='upload-link'>
          Upload
        </router-link>

        <router-link :to="annotationsHref" class="header-item page-link" id='annotations-link'>
          Annotations
        </router-link>

        <router-link :to="datasetsHref" class="header-item page-link" id='datasets-link'>
          Datasets
        </router-link>

        <router-link to="/projects" class="header-item page-link">
          Projects
        </router-link>

        <router-link
          v-if="currentUser && currentUser.primaryGroup"
          :to="primaryGroupHref"
          class="header-item page-link">
          <div class="limit-width">
            {{currentUser.primaryGroup.group.shortName}}
          </div>
        </router-link>
      </div>

      <div class="header-items">
        <router-link to="/help" class="header-item page-link">
          Help
        </router-link>

        <div v-if="loadingUser === 0 && currentUser == null" class="header-items">
          <div class="header-item page-link" @click="showCreateAccount">
            Create account
          </div>

          <div class="header-item page-link" @click="showSignIn">
            Sign in
          </div>
        </div>

        <div v-if="loadingUser === 0 && currentUser != null" class="header-items">
          <div class="submenu-container user-submenu"
               :class="{'submenu-container-open': openSubmenu === 'user'}"
               @mouseenter="handleSubmenuEnter('user')"
               @mouseleave="handleSubmenuLeave('user')">
            <div class="header-item submenu-header"
                 :class="{'router-link-active': matchesRoute('/user/me')}">
              <div class="limit-width" style="color: white;">
              {{ userNameOrEmail }}
              </div>
            </div>
            <div class="submenu">
              <router-link to="/user/me" class="submenu-item page-link">
                My account
              </router-link>
              <div class="submenu-item page-link" @click="logout">
                Sign out
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <el-row v-if="healthMessage" class="alert">
      <el-alert show-icon :title="healthMessage" :type="healthSeverity" :closable="false" />
    </el-row>
  </div>
</template>

<script>
  import gql from 'graphql-tag';
 import {signOut} from '../../api/auth';
  import {getSystemHealthQuery, getSystemHealthSubscribeToMore} from '../../api/system';
 import {encodeParams} from '../Filters';
  import {refreshLoginStatus} from '../../graphqlClient';

 export default {
   name: 'metaspace-header',

   computed: {
     uploadHref() {
       return this.href('/upload');
     },

     datasetsHref() {
       return this.href('/datasets');
     },

     annotationsHref() {
       return this.href('/annotations');
     },

     primaryGroupHref() {
       if (this.currentUser && this.currentUser.primaryGroup) {
         const { id, urlSlug } = this.currentUser.primaryGroup.group;
         return {
           name: 'group',
           params: { groupIdOrSlug: urlSlug || id }
         }
       }
     },

     userNameOrEmail() {
       if (this.currentUser && this.currentUser.name) {
         return this.currentUser.name;
       }
       return '';
     },

     healthMessage() {
       const {canMutate = true, message = null} = this.systemHealth || {};
       if (message) {
         return message;
       } else if (!canMutate) {
         return "METASPACE is currently in read-only mode for scheduled maintenance."
       }
     },
     healthSeverity() {
       return this.systemHealth && this.systemHealth.canMutate === false ? 'warning' : 'info';
     }
   },

   data() {
     return {
       loginEmail: '',
       loadingUser: 0,
       currentUser: null,
       systemHealth: null,
       openSubmenu: null
     };
   },

   apollo: {
     systemHealth: {
       query: getSystemHealthQuery,
       subscribeToMore: getSystemHealthSubscribeToMore,
       fetchPolicy: 'cache-first',
     },
     currentUser: {
       query: gql`query metaspaceHeaderCurrentUserQuery {
         currentUser {
           id
           name
           primaryGroup {
             group {
               id
               shortName
               name
               urlSlug
             }
           }
         }
       }`,
       fetchPolicy: 'cache-first',
       loadingKey: 'loadingUser'
     }
   },

   methods: {
     href(path) {
       const lastParams = this.$store.state.lastUsedFilters[path];
       let f = lastParams ? lastParams.filter : {}
       f = Object.assign({}, f, this.$store.getters.filter)
       const link = {
         path,
         query: encodeParams(f, path, this.$store.state.filterLists)
       };
       return link;
     },

     matchesRoute(path) {
       // WORKAROUND: vue-router hides its util function "isIncludedRoute", which would be perfect here
       // return isIncludedRoute(this.$route, path);
       return this.$route.path.startsWith(path);
     },

     showCreateAccount() {
       this.$store.commit('account/showDialog', 'createAccount');
     },

     showSignIn() {
       this.$store.commit('account/showDialog', 'signIn');
     },

     async logout() {
       await signOut();
       await refreshLoginStatus();
     },

     handleSubmenuEnter(submenu) {
       this.openSubmenu = submenu;
     },
     handleSubmenuLeave(submenu) {
       if (this.openSubmenu === submenu) {
         this.openSubmenu = null;
       }
     },
   }
 }
</script>

<style lang="scss" scoped>
  $header-height: 62px;
  $alert-height: 36px;

 .b-header {
   background-color: rgba(0, 105, 224, 0.85);
   position: fixed;
   z-index: 1000;
   top: 0;
   left: 0;
   right: 0;
   height: 62px;
   display: flex;
   align-items: center;
   justify-content: space-between;
 }

 .spacer {
   height: $header-height + 8px;
 }

 .spacerWithAlert {
   height: $header-height + $alert-height + 8px;
 }

 .header-items {
   display: flex;
   align-items: center;
   height: 100%
 }

 .header-item {
   display: flex;
   border: none;
   padding: 0px 20px;
   font-size: 16px;
   align-self: stretch;
   align-items: center;
   justify-content: center;
 }
 .header-item.logo {
   padding-left: 15px;
 }

 @media (max-width: 1279px) {
   .header-item {
     padding: 0px 10px;
     font-size: 14px;
   }
 }

 .page-link {
   text-align: center;
   color: #eee;
   cursor: pointer;
   text-decoration: none;
 }

 .router-link-active.page-link, .page-link:hover,
 .submenu-container:not(.submenu-container-open) > .router-link-active.submenu-header {
   background: rgba(0, 0, 0, 0.1);
   outline-offset: -1px;
   outline-color: rgba(0, 0, 0, 0.3);
   outline-style: solid;
   outline-width: 1px;
   color: white;
 }

 .router-link-active.page-link {
   font-weight: 700;
 }
 .page-link a {
   text-decoration: none;
 }

 .submenu-container {
   position: relative;
   display: flex;
   align-items: center;
   height: 100%
 }
 .submenu {
   display: none;
   position: absolute;
   flex-direction: column;
   top: 100%;
   right: 0;
   width: 100%;
   min-width: 140px;
   max-width: 200px;

   background-color: rgba(0, 105, 224, 0.85);
   /*background-color: rgb(38, 128, 229);*/
 }
 .submenu-container-open > .submenu {
   display: flex;
 }
 .submenu-item {
   padding: 20px;
   font-size: 16px;
   align-self: stretch;
   justify-content: center;
 }
 .limit-width {
   max-width: 250px;
   overflow-wrap: break-word;
   text-align: center;
   overflow: hidden;
   line-height: 1.2em;
   max-height: 2.4em;
 }
 @media (max-width: 1279px) {
   .limit-width {
     max-width: 150px;
   }
 }

 #email-link-container {
   display: inline-flex;
 }

 .alert {
   position: fixed;
   top: $header-height;
   left: 0;
   right: 0;
   border-radius: 0;
   z-index: 1000;

   .el-alert {
     height: $alert-height;
     justify-content: center;
   }
 }
</style>
