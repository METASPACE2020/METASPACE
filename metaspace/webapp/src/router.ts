/* eslint-disable vue/max-len */
import Vue, { AsyncComponent } from 'vue'
import VueRouter, { RawLocation } from 'vue-router'
import AboutPage from './modules/App/AboutPage.vue'
import DatasetsPage from './modules/Datasets/DatasetsPage.vue'
import { DialogPage, ResetPasswordPage } from './modules/Account'
import { redirectAfterSignIn } from './modules/Account/signInReturnUrl'
import { updateDBParam } from './modules/Filters/url'
import NotFoundPage from './modules/App/NotFoundPage.vue'

Vue.use(VueRouter)

const asyncPagesFreelyTyped = {
  AnnotationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "AnnotationsPage" */ './modules/Annotations/AnnotationsPage.vue'),
  DatasetSummary: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetSummary" */ './modules/Datasets/summary/DatasetSummary.vue'),
  MetadataEditPage: () => import(/* webpackPrefetch: true, webpackChunkName: "MetadataEditPage" */ './modules/MetadataEditor/MetadataEditPage.vue'),
  ImageAlignmentPage: () => import(/* webpackPrefetch: true, webpackChunkName: "ImageAlignmentPage" */ './modules/ImageAlignment/ImageAlignmentPage.vue'),
  UploadPage: () => import(/* webpackPrefetch: true, webpackChunkName: "UploadPage" */ './modules/MetadataEditor/UploadPage.vue'),
  DesignPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DesignPage" */ './design/DesignPage.vue'),

  // These pages are relatively small as they don't have any big 3rd party dependencies, so pack them together
  DatasetTable: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Datasets/list/DatasetTable.vue'),
  HelpPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/HelpPage.vue'),
  EditUserPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/UserProfile/EditUserPage.vue'),
  CreateGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/GroupProfile/CreateGroupPage.vue'),
  ProjectsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Project/ProjectsListPage.vue'),
  SystemHealthPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Admin/SystemHealthPage.vue'),
  GroupsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Admin/GroupsListPage.vue'),
  PrivacyPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/PrivacyPage.vue'),
  TermsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/TermsPage.vue'),
  PublicationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/PublicationsPage.vue'),

  // These pages use sanitizeHtml, which is big
  ViewGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ './modules/GroupProfile/ViewGroupPage.vue'),
  ViewProjectPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ './modules/Project/ViewProjectPage.vue'),
}
const asyncPages = asyncPagesFreelyTyped as Record<keyof typeof asyncPagesFreelyTyped, AsyncComponent>

const convertLegacyUrls = () => {
  const { pathname, hash, search } = window.location
  if (pathname === '/' && hash && hash.startsWith('#/')) {
    history.replaceState(undefined, undefined as any, hash.slice(1))
  }
  if (pathname === '/annotations') {
    const updatedQueryString = updateDBParam(search.slice(1))
    if (updatedQueryString !== null) {
      history.replaceState(undefined, undefined as any, `${pathname}?${updatedQueryString}`)
    }
  }
}
convertLegacyUrls()

const router = new VueRouter({
  mode: 'history',
  routes: [
    { path: '/(about)?', component: AboutPage, meta: { footer: true, headerClass: 'bg-primary' } },
    { path: '/annotations', component: asyncPages.AnnotationsPage },
    {
      path: '/datasets',
      component: DatasetsPage,
      children: [
        { path: '', component: asyncPages.DatasetTable },
        { path: 'summary', component: asyncPages.DatasetSummary },
      ],
    },
    { path: '/datasets/edit/:dataset_id', name: 'edit-metadata', component: asyncPages.MetadataEditPage },
    { path: '/datasets/:dataset_id/add-optical-image', name: 'add-optical-image', component: asyncPages.ImageAlignmentPage },
    { path: '/upload', component: asyncPages.UploadPage },
    { path: '/help', component: asyncPages.HelpPage, meta: { footer: true } },
    { path: '/user/me', component: asyncPages.EditUserPage },

    { path: '/admin/health', component: asyncPages.SystemHealthPage },
    { path: '/admin/groups', component: asyncPages.GroupsListPage },

    { path: '/account/sign-in', component: DialogPage, props: { dialog: 'signIn' } },
    { path: '/account/sign-in-success', redirect: redirectAfterSignIn },
    { path: '/account/create-account', component: DialogPage, props: { dialog: 'createAccount' } },
    { path: '/account/forgot-password', component: DialogPage, props: { dialog: 'forgotPassword' } },
    { path: '/account/reset-password', component: ResetPasswordPage },

    { path: '/group/create', component: asyncPages.CreateGroupPage },
    { path: '/group/:groupIdOrSlug', name: 'group', component: asyncPages.ViewGroupPage },

    { path: '/project/:projectIdOrSlug', name: 'project', component: asyncPages.ViewProjectPage },
    { // Legacy URL sent in "request access" emails up until Feb 2019
      path: '/project/:projectIdOrSlug/manage',
      redirect: { path: '/project/:projectIdOrSlug', query: { tab: 'members' } } as RawLocation,
    },
    { path: '/projects', component: asyncPages.ProjectsListPage },

    { path: '/terms', component: asyncPages.TermsPage, meta: { footer: true } },
    { path: '/privacy', component: asyncPages.PrivacyPage, meta: { footer: true } },
    { path: '/publications', component: asyncPages.PublicationsPage, meta: { footer: true } },

    { path: '/design', component: asyncPages.DesignPage, meta: { footer: true, flex: true } },

    { path: '*', component: NotFoundPage, meta: { footer: true, flex: true } },
  ],
  scrollBehavior(to, from, savedPosition) {
    if (to.hash) {
      return { selector: to.hash, offset: { x: 0, y: 64 } } // offset header height
    }
    if (savedPosition) {
      return savedPosition
    }
    if (to.path !== from.path) {
      return { x: 0, y: 0 }
    }
  },
})

const { href } = router.resolve({ name: 'project', params: { projectIdOrSlug: 'REMOVE' } }, undefined, true)
export const PROJECT_URL_PREFIX = location.origin + href.replace('REMOVE', '')

export default router
