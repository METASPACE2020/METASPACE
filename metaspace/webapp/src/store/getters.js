import {mzFilterPrecision} from '../util';
import {decodeParams, decodeSettings, getLevel} from '../modules/Filters';
import config from '../config';

/** For filters where empty string is a valid client-side value that must be converted to empty string for the API */
const noneToEmptyString = s => s === 'none' ? '' : s;

export default {
  filterLevel(state) {
    return getLevel(state.route.path);
  },

  filter(state) {
    return decodeParams(state.route, state.filterLists);
  },

  settings(state) {
    return decodeSettings(state.route);
  },

  ftsQuery(state, getters) {
    return getters.filter.simpleQuery;
  },

  gqlAnnotationFilter(state, getters) {
    const filter = getters.filter;
    const colocalizationAlgo = getters.settings.annotationView.colocalizationAlgo;

    const f = {
      database: filter.database,
      compoundQuery: filter.compoundName,
      chemMod: noneToEmptyString(filter.chemMod),
      neutralLoss: noneToEmptyString(filter.neutralLoss),
      adduct: filter.adduct,
      fdrLevel: filter.fdrLevel,
      colocalizedWith: filter.colocalizedWith,
      // Only include colocalizationAlgo if there is another filter that uses it. Otherwise the annotations list
      // refreshes unnecessarily when changing algorithm.
      colocalizationAlgo: filter.colocalizedWith || filter.colocalizationSamples ? colocalizationAlgo : null,
      colocalizationSamples: filter.colocalizationSamples,
      offSample: filter.offSample == null ? undefined : !!filter.offSample,
    };

    if (!config.features.neutral_losses && !config.features.advanced_ds_config) {
      f.hasNeutralLoss = false;
    }

    if (!config.features.chem_mods && !config.features.advanced_ds_config) {
      f.hasChemMod = false;
    }

    if (!config.features.all_adducts) {
      f.hasHiddenAdduct = false;
    }

    if (filter.minMSM)
      f.msmScoreFilter = {min: filter.minMSM, max: 1.0};

    if (filter.mz) {
      const mz = parseFloat(filter.mz),
            deltamz = parseFloat(mzFilterPrecision(mz));
      f.mzFilter = {
        min: mz - deltamz,
        max: mz + deltamz
      };
    }

    return f;
  },

  gqlDatasetFilter(state, getters) {
    const filter = getters.filter;
    const {group, project, submitter, datasetIds, polarity,
           organism, organismPart, condition, growthConditions,
           ionisationSource, analyzerType, maldiMatrix, metadataType,
           compoundName} = filter;
    const level = getters.filterLevel;
    const hasAnnotationMatching = level === 'dataset' && compoundName ? { compoundQuery: compoundName } : undefined;
    return {
      group: group,
      project: project,
      submitter: submitter,

      // temporary workaround because of array-related bugs in apollo-client
      ids: datasetIds ? datasetIds.join("|") : null,

      organism,
      organismPart,
      condition,
      growthConditions,
      ionisationSource,
      maldiMatrix,
      analyzerType,
      polarity: polarity ? polarity.toUpperCase() : null,
      metadataType,
      hasAnnotationMatching,
    }
  },

  gqlColocalizationFilter(state, getters) {
    const {datasetIds, colocalizedWith, database, fdrLevel} = getters.filter;
    const colocalizationAlgo = getters.settings.annotationView.colocalizationAlgo;
    if (datasetIds && !datasetIds.includes('|') && colocalizedWith != null && database != null && fdrLevel != null) {
      return {colocalizedWith, colocalizationAlgo, database, fdrLevel};
    } else {
      return null;
    }
  },
}
