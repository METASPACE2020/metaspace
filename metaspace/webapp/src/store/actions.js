import { fetchOptionListsQuery } from '../api/metadata'
import { decodeParams } from '../modules/Filters'
import { computed } from 'vue'
import store from '../store/index'

export default {
  async initFilterLists(context, apolloClient) {
    if (context.state.filterListsLoading || context.state.filterLists != null) return
    await context.commit('setFilterListsLoading')

    const response = await apolloClient.query({
      query: fetchOptionListsQuery,
      fetchPolicy: 'cache-first',
    })

    // set annotationIds default according to values passed from snapshot to the store state
    context.commit('setFilterLists', {
      ...response.data,
      annotationIds: computed(() => store.state.snapshotAnnotationIds),
    })

    // Refresh the current filter so that computed defaults that depend on `filterLists` are applied
    const filter = decodeParams(context.state.route, context.state.filterLists)

    context.commit('updateFilter', filter)
  },
}
