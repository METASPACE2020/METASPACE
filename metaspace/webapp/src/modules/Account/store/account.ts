import { Module } from 'vuex';
import { DialogType } from '../dialogs';
import router from '../../../router';

export interface AccountState {
  dialog: DialogType | null;
  showDialogAsPage: boolean;
}

const account = {
  namespaced: true,
  state: {
    dialog: null,
    showDialogAsPage: false
  },
  mutations: {
    showDialog(state: AccountState, dialog: DialogType) {
      state.dialog = dialog;
      state.showDialogAsPage = false;
    },
    showDialogAsPage(state: AccountState, dialog: DialogType) {
      state.dialog = dialog;
      state.showDialogAsPage = true;
    },
    hideDialog(state: AccountState, dialog: DialogType) {
      if (state.dialog === dialog) {
        state.dialog = null;
        if (state.showDialogAsPage) {
          router.push('/');
        }
      }
    }
  }
};

export default account as Module<AccountState, any>;
