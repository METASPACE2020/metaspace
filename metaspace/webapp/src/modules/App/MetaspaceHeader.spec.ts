import { nextTick, ref } from 'vue'
import router from '../../router'
import store from '../../store'
import { mount } from '@vue/test-utils'
import MetaspaceHeader from './MetaspaceHeader'
import { expect } from 'vitest'

const currentUserMockResponses = [
  {
    data: {
      currentUser: {
        id: '123',
        name: 'Test UserX',
        email: 'test@example.com',
        primaryGroup: {
          group: {
            id: '456',
            name: 'Test Group',
            urlSlug: null,
          },
        },
        groups: [
          { role: 'MEMBER', group: { hasPendingRequest: null } },
          { role: 'GROUP_ADMIN', group: { hasPendingRequest: false } },
        ],
        projects: [
          { role: 'PENDING', project: { hasPendingRequest: null } },
          { role: 'MANAGER', project: { hasPendingRequest: true } },
        ],
      },
    },
  },
  {
    data: {
      currentUser: null,
    },
  },
]

let mockIndex = 0 // This will determine which mock response to use

vi.mock('@vue/apollo-composable', () => ({
  useQuery: vi.fn(() => ({
    result: ref(currentUserMockResponses[mockIndex].data),
    loading: ref(false),
    error: ref(null),
    subscribeToMore: vi.fn(),
  })),
}))

// Mock router's push function or any other function you need
router.push = vi.fn()

describe('MetaspaceHeader', () => {
  it('should match snapshot (logged in)', async () => {
    mockIndex = 0
    const wrapper = mount(MetaspaceHeader, {
      global: {
        plugins: [router, store],
      },
    })

    await nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should match snapshot (logged out)', async () => {
    mockIndex = 1
    const wrapper = mount(MetaspaceHeader, {
      global: {
        plugins: [router, store],
      },
    })

    await nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })
})
