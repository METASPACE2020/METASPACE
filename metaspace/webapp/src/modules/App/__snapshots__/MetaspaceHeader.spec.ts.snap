// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetaspaceHeader should match snapshot (logged in) 1`] = `
<div class="b-header">
  <div class="header-items">
    <a href="/" class="header-item logo router-link-active">
      <img src="../../assets/logo.png" alt="Metaspace" title="Metaspace">
    </a>
    <a href="/upload" class="header-item page-link" id="upload-link">
      Upload
    </a>
    <a href="/annotations" class="header-item page-link" id="annotations-link">
      Annotations
    </a>
    <a href="/datasets" class="header-item page-link" id="datasets-link">
      Datasets
    </a>
    <a href="/projects" class="header-item page-link">
      Projects
    </a>
    <a href="/group/456" class="header-item page-link">
      <div class="limit-width">
        currentUser.primaryGroup.group.shortName
      </div>
    </a>
  </div>
  <div class="header-items">
    <a href="/help" class="header-item page-link">
      Help
    </a>
    <!---->
    <div class="header-items">
      <div class="submenu-container user-submenu">
        <div class="header-item submenu-header">
          <div class="limit-width" style="color: white;">
            Test User
          </div>
        </div>
        <div class="submenu">
          <a href="/user/me" class="submenu-item page-link">
            My account
          </a>
          <div class="submenu-item page-link">
            Sign out
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MetaspaceHeader should match snapshot (logged out) 1`] = `
<div class="b-header">
  <div class="header-items">
    <a href="/" class="header-item logo router-link-active">
      <img src="../../assets/logo.png" alt="Metaspace" title="Metaspace">
    </a>
    <a href="/upload" class="header-item page-link" id="upload-link">
      Upload
    </a>
    <a href="/annotations" class="header-item page-link" id="annotations-link">
      Annotations
    </a>
    <a href="/datasets" class="header-item page-link" id="datasets-link">
      Datasets
    </a>
    <a href="/projects" class="header-item page-link">
      Projects
    </a>
    <!---->
  </div>
  <div class="header-items">
    <a href="/help" class="header-item page-link">
      Help
    </a>
    <div class="header-items">
      <div class="header-item page-link">
        Create account
      </div>
      <div class="header-item page-link">
        Sign in
      </div>
    </div>
    <!---->
  </div>
</div>
`;
